### ISSUE TITLE
Support iterating publishables from Jervis to Jenkins

### ISSUE THREAD

**Issue by samrocketman** [samrocketman/jervis#124]
Currently, it is quite complex for an admin to define collections for publishing.  This is mainly due to it being complicated to define settings and override them.

<details><summary>Complicated Ugly Example (click to expand)</summary>

```groovy
def processDefaultPublishable(def item, String publishable, boolean is_pull_request) {
    if(is_pull_request && ('skip_on_pr' in item) && (item['skip_on_pr'] in Boolean) && item['skip_on_pr']) {
        echo "Skip publishing ${publishable} for pull request."
        //skip because we shouldn't publish on a pull request
        return
    }
    switch(publishable) {
        case 'artifacts':
            archiveArtifacts artifacts: item['path'], fingerprint: true,
                             excludes: item['excludes'],
                             allowEmptyArchive: item['allowEmptyArchive'],
                             defaultExcludes: item['defaultExcludes'],
                             caseSensitive: item['caseSensitive']
            break
        case 'cobertura':
            cobertura coberturaReportFile: item['path'],
                      autoUpdateHealth: item['autoUpdateHealth'],
                      autoUpdateStability: item['autoUpdateStability'],
                      failNoReports: item['failNoReports'],
                      failUnhealthy: item['failUnhealthy'],
                      failUnstable: item['failUnstable'],
                      maxNumberOfBuilds: item['maxNumberOfBuilds'],
                      onlyStable: item['onlyStable'],
                      sourceEncoding: item['sourceEncoding'],
                      zoomCoverageChart: item['zoomCoverageChart'],
                      methodCoverageTargets: item['methodCoverageTargets'],
                      lineCoverageTargets: item['lineCoverageTargets'],
                      conditionalCoverageTargets: item['conditionalCoverageTargets']
            break
        case 'html':
            publishHTML allowMissing: item['allowMissing'],
                        alwaysLinkToLastBuild: item['alwaysLinkToLastBuild'],
                        includes: item['includes'],
                        keepAll: item['keepAll'],
                        reportDir: item['path'],
                        reportFiles: item['reportFiles'],
                        reportName: item['reportName'],
                        reportTitles: item['reportTitles']
            break
        case 'junit':
            junit allowEmptyResults: item['allowEmptyResults'],
                  healthScaleFactor: item['healthScaleFactor'],
                  keepLongStdio: item['keepLongStdio'],
                  testResults: item['path']
            break
    }
}
```

Which is used in

```groovy
            stage("Publish results") {
                //unstash and publish in parallel
                Map tasks = [failFast: true]
                for(String publishable : publishableItems) {
                    String publish = publishable
                    tasks["Publish ${publish}"] = {
                        try {
                            unstash publish
                            processDefaultPublishable(pipeline_generator.getPublishable(publish), publish, is_pull_request)
                        }
                        catch(e) {
                            currentBuild.result = 'FAILURE'
                        }
                    }
                }
                parallel(tasks)
            }

```

</details><br />

Since the `default_collect_settings` map closely match the map being passed to the Jenkins publisher then we should be able to define collections in a programmatic way.

# JUnit Example

We can call methods from Strings in Groovy.  Additionally, the current collection is a Groovy shorthand for calling a method with a Map.

```groovy
junit allowEmptyResults: item['allowEmptyResults'],
      healthScaleFactor: item['healthScaleFactor'],
      keepLongStdio: item['keepLongStdio'],
      testResults: item['path']
```

Is the same as:

```groovy
Map settings = [
    allowEmptyResults: item['allowEmptyResults'],
    healthScaleFactor: item['healthScaleFactor'],
    keepLongStdio: item['keepLongStdio'],
    testResults: item['path']
]
junit(settings)
```

Which is also the same as:

```groovy
String method = 'junit'
Map settings = [
    allowEmptyResults: item['allowEmptyResults'],
    healthScaleFactor: item['healthScaleFactor'],
    keepLongStdio: item['keepLongStdio'],
    testResults: item['path']
]
"${method}"(settings)
```

# Things to account for

The settings map does not exactly match the map for settings when calling a publisher in Jenkins.  `skip_on_pr` is used from settings in the default YAML map but is not used by the Jenkins publisher.  I'll need to support removing this item automatically from settings.

Example from JUnit, `testResults` is coming from `path`.  I'll need to support translating from `path` to `testResults` in the iterator.

The method name used to publish items does not always match the publisher in Jervis settings.  For example,

```yaml
jenkins:
  collect:
    artifacts: ...
```

Uses the `archiveArtifacts` method name to publish artifacts in Jenkins.  So I'll need the ability to define alternate method names where methods do not match the Jervis collector.

# Feature behavior

So finally let's assume all of the above is accounted for.  Pipeline DSL _could look_ something like this instead of the ugly DSL example presented at the beginning of this issue.

```groovy
for(String publishable : pipeline_generator.publishableItems) {
    //necessary for using variable with unstash in parallel steps; otherwise it won't work with a for loop
    String publish = publishable
    if(is_pull_request && pipeline_generator.skipPublishableOnPR(publish)) {
        continue
    }
    try {
        unstash publish
        //call the method from String with settings using Groovy fanciness
        "${pipeline_generator.getPublishableMethodName(publish)}"( pipeline_generator.getPublishableMethodSettings(publish) )
    }
    catch(e) {
        currentBuild.result = 'FAILURE'
    }
}
```

**Comment by samrocketman** [comment]
Won't do.
